src/data/s3-data-source.ts

```diff
@@ -0,0 +1,115 @@
+/**
+ * AR.IO Gateway
+ * Copyright (C) 2022-2023 Permanent Data Solutions, Inc. All Rights Reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import winston from 'winston';
+import { headerNames } from '../constants.js';
+
+import {
+  ContiguousData,
+  ContiguousDataSource,
+  RequestAttributes,
+} from '../types.js';
+import { AwsLiteS3 } from '@aws-lite/s3-types';
+import { Readable } from 'node:stream';
+
+export class S3DataSource implements ContiguousDataSource {
+  private log: winston.Logger;
+  private s3Client: AwsLiteS3;
+  private s3Bucket: string;
+  private s3Prefix: string;
+
+  constructor({
+    log,
+    s3Client,
+    s3Bucket,
+    s3Prefix = '',
+  }: {
+    log: winston.Logger;
+    s3Client: AwsLiteS3;
+    s3Bucket: string;
+    s3Prefix?: string;
+  }) {
+    this.log = log.child({ class: 'S3DataSource' });
+    this.s3Client = s3Client;
+    this.s3Bucket = s3Bucket;
+    this.s3Prefix = s3Prefix;
+  }
+
+  async getData({
+    id,
+    requestAttributes,
+  }: {
+    id: string;
+    requestAttributes?: RequestAttributes;
+  }): Promise<ContiguousData> {
+    const log = this.log.child({ method: 'getData' });
+    log.info('Fetching contiguous data from S3', {
+      id,
+      bucket: this.s3Bucket,
+      prefix: this.s3Prefix,
```

Comment:
Could it be useful to see the requestAttributes here as well? Also do we want these to be info or debug logs?

Reply:
`requestAttributes` are just `hops` and `origin` which are not propagated to S3.

In this case we just increment `hops` and return the new value.